Provide fidelity bounds as output of MPS-based simulations

AlbertoBaiardi opened the pull request:
### Summary

This PR implements the calculation of the fidelity bounds on MPS-based simulations.

### Details and comments

The milestones are the following:

 - [x] implement the fidelity bounds for the Vidal canonization + corresponding tests
 - [x] implement the fidelity bounds for the conventional canonization + corresponding tests
 - [ ] implement the fidelity bounds for HOBOs

eggerdj commented:
Regarding: https://github.com/qiskit-community/qaoa_training_pipeline/pull/6#issuecomment-3096274175 sounds good to me. Let's keep it seperate.

eggerdj reviewed (commented):
This looks good to me. I made a few suggestions. One last point, does this slow the MPSEvaluator down at all? Even when we don't store the Schmidt values?

AlbertoBaiardi commented on qaoa_training_pipeline/evaluation/mps_evaluator.py:
```diff
@@ -84,14 +87,18 @@ def __init__(
                 network. Defaults to False.
             store_schmidt_values (bool): If true then the Schmidt values are stored in a JSON
                 serializable format.
+            store_intermediate_schmidt_values (bool): If `True`, stores the intermediate Schmidt
+                values obtained at each application of a two-qubit gate.
```
True -- fixed in a0fa77372501e1307d01c6dd81df0d2f5f66fc0e

AlbertoBaiardi commented on qaoa_training_pipeline/evaluation/mps_evaluator.py:
```diff
@@ -231,6 +242,64 @@ def get_results_from_last_iteration(self) -> Dict:
         """
         return self._results_last_iteration
 
+    def calculate_fidelity_bounds(self) -> float:
```
Totally -- I fixed this in bcd7e0bdfdc4508219c93ecf90d90a75a8ca5603 (I also updated the corresponding tests)

AlbertoBaiardi commented on qaoa_training_pipeline/utils/tns_utils/qaoa_circuit_mps.py:
```diff
@@ -615,28 +689,28 @@ def _apply_layer_ansatz_swap_strat(self, scaling_factor: float, rep: int):
 
         # There are len(layer_order) layers of Rzz gates and len(layer_order) - 1 layers
         # of SWAP gates.
+        swap_gate = parse_to_gate("SWAP").build_array()
         for layer_idx in layer_order:
+
             permutation = self._swap_strat.inverse_composed_permutation(layer_idx)
+
             # 1. Apply the gates.
             for i_pairs in self._swap_layer_pairs[layer_idx]:
                 j_qubit, i_qubit = i_pairs[0], i_pairs[1]
                 tn_i = permutation.index(i_qubit)
                 tn_j = permutation.index(j_qubit)
 
-                tn_j_qubit = max(tn_i, tn_j)
+                tn_i_qubit = min(tn_i, tn_j)
```
Here I assumed that `tn_i` and `tn_j` are nearest-neighbour -- is this correct? If this the case, I just took the minimum as this is the site where the canonization will be centered

AlbertoBaiardi commented:
> @AlbertoBaiardi can you also update the version in the system info?
> 
> https://github.com/qiskit-community/qaoa_training_pipeline/blob/e1f4b2e275be3111467584def16694987ed8b73a/qaoa_training_pipeline/training/param_result.py#L49
> this should be accompanied with a corresponding update in the main README.md.

True -- I did not see that as I forgot to update the PR with the newest `main`.

I fixed this in cd5a0968b8fa345355a58bbbecae5360db5caacb and 3206193a201ba3cc03828ecf94d57a8b08e04891 (where I updated also the README file)

AlbertoBaiardi commented on qaoa_training_pipeline/utils/tns_utils/qaoa_circuit_mps.py:
```diff
@@ -321,6 +345,11 @@ def _apply_ansatz_layer(self, scaling_factor: float):
         Args:
             scaling_factor (float): scaling term appearing in the exponent.
+
+        Returns:
```
Yep - done in 33fdb53c6dfc30143fc1c46e321b90042f8d6321
